File: vi/codes/cpp/chapter_hashing/hash_map_open_addressing.cpp
/**
 * File: hash_map_open_addressing.cpp
 * Created Time: 2023-06-13
 * Author: krahets (krahets@163.com)
 */

#include "./array_hash_map.cpp"

/* Bảng băm sử dụng phương pháp địa chỉ mở */
class HashMapOpenAddressing {
  private:
    int size;                             // Số lượng cặp khóa-giá trị
    int capacity = 4;                     // Dung lượng bảng băm
    const double loadThres = 2.0 / 3.0;     // Ngưỡng hệ số tải để kích hoạt mở rộng
    const int extendRatio = 2;            // Bội số mở rộng
    vector<Pair *> buckets;               // Mảng bucket (ngăn chứa)
    Pair *TOMBSTONE = new Pair(-1, "-1"); // Dấu hiệu đã xóa

  public:
    /* Hàm khởi tạo */
    HashMapOpenAddressing() : size(0), buckets(capacity, nullptr) {
    }

    /* Hàm hủy */
    ~HashMapOpenAddressing() {
        for (Pair *pair : buckets) {
            if (pair != nullptr && pair != TOMBSTONE) {
                delete pair;
            }
        }
        delete TOMBSTONE;
    }

    /* Hàm băm */
    int hashFunc(int key) {
        return key % capacity;
    }

    /* Hệ số tải */
    double loadFactor() {
        return (double)size / capacity;
    }

    /* Tìm chỉ số bucket tương ứng với key */
    int findBucket(int key) {
        int index = hashFunc(key);
        int firstTombstone = -1;
        // Dò tuyến tính, dừng khi gặp bucket rỗng
        while (buckets[index] != nullptr) {
            // Nếu gặp key, trả về chỉ số bucket tương ứng
            if (buckets[index]->key == key) {
                // Nếu trước đó đã gặp dấu hiệu đã xóa, di chuyển cặp khóa-giá trị đến vị trí đó
                if (firstTombstone != -1) {
                    buckets[firstTombstone] = buckets[index];
                    buckets[index] = TOMBSTONE;
                    return firstTombstone; // Trả về chỉ số bucket đã di chuyển đến
                }
                return index; // Trả về chỉ số bucket
            }
            // Ghi lại dấu hiệu đã xóa đầu tiên gặp được
            if (firstTombstone == -1 && buckets[index] == TOMBSTONE) {
                firstTombstone = index;
            }
            // Tính chỉ số bucket, quay lại đầu nếu vượt quá cuối
            index = (index + 1) % capacity;
        }
        // Nếu key không tồn tại, trả về chỉ số để chèn vào
        return firstTombstone == -1 ? index : firstTombstone;
    }

    /* Thao tác truy vấn */
    string get(int key) {
        // Tìm chỉ số bucket tương ứng với key
        int index = findBucket(key);
        // Nếu tìm thấy cặp khóa-giá trị, trả về val tương ứng
        if (buckets[index] != nullptr && buckets[index] != TOMBSTONE) {
            return buckets[index]->val;
        }
        // Nếu cặp khóa-giá trị không tồn tại, trả về chuỗi rỗng
        return "";
    }

    /* Thao tác thêm */
    void put(int key, string val) {
        // Khi hệ số tải vượt ngưỡng, thực hiện mở rộng
        if (loadFactor() > loadThres) {
            extend();
        }
        // Tìm chỉ số bucket tương ứng với key
        int index = findBucket(key);
        // Nếu tìm thấy cặp khóa-giá trị, ghi đè val và trả về
        if (buckets[index] != nullptr && buckets[index] != TOMBSTONE) {
            buckets[index]->val = val;
            return;
        }
        // Nếu cặp khóa-giá trị không tồn tại, thêm cặp khóa-giá trị
        buckets[index] = new Pair(key, val);
        size++;
    }

    /* Thao tác xóa */
    void remove(int key) {
        // Tìm chỉ số bucket tương ứng với key
        int index = findBucket(key);
        // Nếu tìm thấy cặp khóa-giá trị, ghi đè bằng dấu hiệu đã xóa
        if (buckets[index] != nullptr && buckets[index] != TOMBSTONE) {
            delete buckets[index];
            buckets[index] = TOMBSTONE;
            size--;
        }
    }

    /* Mở rộng bảng băm */
    void extend() {
        // Lưu tạm bảng băm gốc
        vector<Pair *> bucketsTmp = buckets;
        // Khởi tạo bảng băm mới đã mở rộng
        capacity *= extendRatio;
        buckets = vector<Pair *>(capacity, nullptr);
        size = 0;
        // Di chuyển các cặp khóa-giá trị từ bảng băm gốc sang bảng băm mới
        for (Pair *pair : bucketsTmp) {
            if (pair != nullptr && pair != TOMBSTONE) {
                put(pair->key, pair->val);
                delete pair;
            }
        }
    }

    /* In bảng băm */
    void print() {
        for (Pair *pair : buckets) {
            if (pair == nullptr) {
                cout << "nullptr" << endl;
            } else if (pair == TOMBSTONE) {
                cout << "TOMBSTONE" << endl;
            } else {
                cout << pair->key << " -> " << pair->val << endl;
            }
        }
    }
};

/* Driver Code */
int main() {
    // Khởi tạo bảng băm
    HashMapOpenAddressing hashmap;

    // Thao tác thêm
    // Thêm cặp khóa-giá trị (key, val) vào bảng băm
    hashmap.put(12836, "Xiao Ha");
    hashmap.put(15937, "Xiao Luo");
    hashmap.put(16750, "Xiao Suan");
    hashmap.put(13276, "Xiao Fa");
    hashmap.put(10583, "Xiao Ya");
    cout << "\nAfter adding is complete, hash table is\nKey -> Value" << endl;
    hashmap.print();

    // Thao tác truy vấn
    // Đưa khóa vào bảng băm để lấy giá trị val
    string name = hashmap.get(13276);
    cout << "\nInput student ID 13276, query name " << name << endl;

    // Thao tác xóa
    // Xóa cặp khóa-giá trị (key, val) khỏi bảng băm
    hashmap.remove(16750);
    cout << "\nAfter removing 16750, hash table is\nKey -> Value" << endl;
    hashmap.print();

    return 0;
}

File: vi/codes/swift/chapter_hashing/hash_map_open_addressing.swift
/**
 * File: hash_map_open_addressing.swift
 * Created Time: 2023-06-28
 * Author: nuomi1 (nuomi1@qq.com)
 */

import utils

/* Bảng băm dùng phương pháp địa chỉ mở (open addressing) */
class HashMapOpenAddressing {
    var size: Int // Số lượng cặp khóa-giá trị
    var capacity: Int // Dung lượng bảng băm
    var loadThres: Double // Ngưỡng hệ số tải để kích hoạt mở rộng
    var extendRatio: Int // Hệ số mở rộng
    var buckets: [Pair?] // Mảng bucket
    var TOMBSTONE: Pair // Dấu hiệu đã xóa (bia mộ)

    /* Hàm khởi tạo */
    init() {
        size = 0
        capacity = 4
        loadThres = 2.0 / 3.0
        extendRatio = 2
        buckets = Array(repeating: nil, count: capacity)
        TOMBSTONE = Pair(key: -1, val: "-1")
    }

    /* Hàm băm */
    func hashFunc(key: Int) -> Int {
        key % capacity
    }

    /* Hệ số tải */
    func loadFactor() -> Double {
        Double(size) / Double(capacity)
    }

    /* Tìm chỉ số bucket tương ứng với khóa */
    func findBucket(key: Int) -> Int {
        var index = hashFunc(key: key)
        var firstTombstone = -1
        // Dò tuyến tính, dừng lại khi gặp bucket trống
        while buckets[index] != nil {
            // Nếu gặp khóa, trả về chỉ số bucket tương ứng
            if buckets[index]!.key == key {
                // Nếu trước đó đã gặp dấu hiệu đã xóa, di chuyển cặp khóa-giá trị đến vị trí đó
                if firstTombstone != -1 {
                    buckets[firstTombstone] = buckets[index]
                    buckets[index] = TOMBSTONE
                    return firstTombstone // Trả về chỉ số bucket sau khi di chuyển
                }
                return index // Trả về chỉ số bucket
            }
            // Ghi nhận dấu hiệu đã xóa đầu tiên gặp phải
            if firstTombstone == -1 && buckets[index] == TOMBSTONE {
                firstTombstone = index
            }
            // Tính chỉ số bucket tiếp theo, quay về đầu nếu vượt quá cuối
            index = (index + 1) % capacity
        }
        // Nếu khóa không tồn tại, trả về chỉ số để chèn
        return firstTombstone == -1 ? index : firstTombstone
    }

    /* Thao tác truy vấn */
    func get(key: Int) -> String? {
        // Tìm chỉ số bucket tương ứng với khóa
        let index = findBucket(key: key)
        // Nếu tìm thấy cặp khóa-giá trị, trả về val tương ứng
        if buckets[index] != nil, buckets[index] != TOMBSTONE {
            return buckets[index]!.val
        }
        // Nếu cặp khóa-giá trị không tồn tại, trả về null
        return nil
    }

    /* Thao tác thêm */
    func put(key: Int, val: String) {
        // Khi hệ số tải vượt ngưỡng, tiến hành mở rộng
        if loadFactor() > loadThres {
            extend()
        }
        // Tìm chỉ số bucket tương ứng với khóa
        let index = findBucket(key: key)
        // Nếu tìm thấy cặp khóa-giá trị, ghi đè val rồi trả về
        if buckets[index] != nil, buckets[index] != TOMBSTONE {
            buckets[index]!.val = val
            return
        }
        // Nếu cặp khóa-giá trị không tồn tại, thêm cặp khóa-giá trị mới
        buckets[index] = Pair(key: key, val: val)
        size += 1
    }

    /* Thao tác xóa */
    func remove(key: Int) {
        // Tìm chỉ số bucket tương ứng với khóa
        let index = findBucket(key: key)
        // Nếu tìm thấy cặp khóa-giá trị, ghi đè bằng dấu hiệu đã xóa
        if buckets[index] != nil, buckets[index] != TOMBSTONE {
            buckets[index] = TOMBSTONE
            size -= 1
        }
    }

    /* Mở rộng bảng băm */
    func extend() {
        // Lưu tạm bảng băm gốc
        let bucketsTmp = buckets
        // Khởi tạo bảng băm mới đã mở rộng
        capacity *= extendRatio
        buckets = Array(repeating: nil, count: capacity)
        size = 0
        // Chuyển các cặp khóa-giá trị từ bảng băm gốc sang bảng băm mới
        for pair in bucketsTmp {
            if let pair, pair != TOMBSTONE {
                put(key: pair.key, val: pair.val)
            }
        }
    }

    /* In bảng băm */
    func print() {
        for pair in buckets {
            if pair == nil {
                Swift.print("null")
            } else if pair == TOMBSTONE {
                Swift.print("TOMBSTONE")
            } else {
                Swift.print("\(pair!.key) -> \(pair!.val)")
            }
        }
    }
}

@main
enum _HashMapOpenAddressing {
    /* Mã điều khiển */
    static func main() {
        /* Khởi tạo bảng băm */
        let map = HashMapOpenAddressing()

        /* Thao tác thêm */
        // Thêm cặp khóa-giá trị (key, value) vào bảng băm
        map.put(key: 12836, val: "Xiao Ha")
        map.put(key: 15937, val: "Xiao Luo")
        map.put(key: 16750, val: "Xiao Suan")
        map.put(key: 13276, val: "Xiao Fa")
        map.put(key: 10583, val: "Xiao Ya")
        print("\nAfter adding is complete, hash table is\nKey -> Value")
        map.print()

        /* Thao tác truy vấn */
        // Đưa khóa vào bảng băm để lấy giá trị
        let name = map.get(key: 13276)
        print("\nInput student ID 13276, found name \(name!)")

        /* Thao tác xóa */
        // Xóa cặp khóa-giá trị (key, value) khỏi bảng băm
        map.remove(key: 16750)
        print("\nAfter removing 16750, hash table is\nKey -> Value")
        map.print()
    }
}
